### PR TITLE
CMake build for S2S: Successful Compile/Build

### PR DESCRIPTION
**Pull request by BrianCurtis-NOAA**
### Description of changes
Brought over CMake from ufs-weather-model feature/s2s_cmake
Fixes for BUILD_INTERFACE
Cleaned up CMakeLists.txt between -weather and -s2s. Added script to … 
rename build.sh to build_cmake.sh for a handy build test script, crea…
Typo in OpenMP CMake call
Edit compile_cmake to include MOM6/CICE6/CMEPS build options and use …
Added FindPIO.cmake to transition CICE from its internal PIO build 
Fix for -qopenmp call using MOM6, bringing tests_cmake dir for CMake … 
Fixes for using tests_cmake dir
CICE CMake fixes
NEMS.exe include
MOM6 had extra file, cleaned up to match MOM6 GMake flow better, test… 
Don't need fv3_stack now that fv3 has stack as default
ESMF VERSION MAJOR/MINOR missing from CMEPS
Fixed merge conflict from upstream: updated modulefiles fv3 to include pio
Matching GMake compiler flags in CMake
Cleanup of CICE/MOM6 CMakeLists.txt 

### Specific notes
To JUST build with CMake:
`module use modulefiles/orion.intel`
`module load fv3`
`./build_cmake.sh`
-This PR does not address regression testing.

Issues Fixed (include github issue #):
#94 #95 

Are changes expected to change answers?
- bit for bit answers should NOT change
- more substantial answers should NOT change

Specific changes:
- changes in parm directory: No.
- changes in module files: Yes. Creation of fv3_stack for orion.intel to use hpc-stack
- new tests added or removed: Yes. Specific test files for S2S (will unify with -weather later)

Testing performed:
- machines: orion, hera
- This PR is not meant to address regression testing yet.

Hashes used for testing:
- NEMS: 9d05172
- CMEPS: 5e63bd4
- FV3: 0975bb6
- MOM6: eb58a2e
- CICE6: 285985c
- WW3: 96e3f3a
- FMS: f61416f
- stochastic_physics: a8e2cc1 

Co-authored by:
@aerorahul

### CODE REVIEW

File: MOM6-interface/CMakeLists.txt [hunk 1]
```diff
@@ -7,38 +7,44 @@ list(APPEND lib_src_files
...
 
 target_include_directories(${libName} INTERFACE
   $<BUILD_INTERFACE:${moduleDir}>
   $<INSTALL_INTERFACE:include/${libName}>)
 
-if(OpenMP_Fortran_FOUND)
+target_link_libraries(${libName} PUBLIC fms)
+target_link_libraries(${libName} PUBLIC NetCDF::NetCDF_Fortran)
+target_link_libraries(${libName} PUBLIC esmf)
+if(OpenMP)
```
Comment (kgerheiser) [inline MOM6-interface/CMakeLists.txt]: Should this be `OPENMP` (all caps)?

Comment (BrianCurtis-NOAA) [reply MOM6-interface/CMakeLists.txt]: Yes! Thanks.

File: CMakeLists.txt [hunk 2]
```diff
@@ -122,22 +155,70 @@ if(CCPP)
...
+
+target_link_libraries(fms PRIVATE NetCDF::NetCDF_Fortran)
+set_target_properties(fms PROPERTIES ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/FMS)
+set_target_properties(fms PROPERTIES Fortran_MODULE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/FMS/mod)
+target_include_directories(fms PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/FMS/mod>)
+
+###############################################################################
+### CMEPS
+###############################################################################
+if(CMEPS)
```
Comment (aerorahul) [inline CMakeLists.txt]: why do you need an if-block in S2S model.  Isn't that mandatory here?

File: CMakeLists.txt [hunk 3]
```diff
@@ -122,22 +155,70 @@ if(CCPP)
...
 ### FV3
 ###############################################################################
 add_subdirectory(FV3)
 
 ###############################################################################
-### MOM6
+### CICE_interface
 ###############################################################################
-add_subdirectory(MOM6-interface)
+if(CICE6)
```
Comment (aerorahul) [inline CMakeLists.txt]: Why if-block in S2S?

File: CMakeLists.txt [hunk 4]
```diff
@@ -122,22 +155,70 @@ if(CCPP)
...
+    add_subdirectory(CICE-interface)
+endif()
 
 ###############################################################################
-### CICE
+### MOM6-interface
 ###############################################################################
-add_subdirectory(CICE-interface)
+if(MOM6)
```
Comment (aerorahul) [inline CMakeLists.txt]: same here

File: CMakeLists.txt [hunk 5]
```diff
@@ -122,22 +155,70 @@ if(CCPP)
...
+    add_subdirectory(MOM6-interface)
+endif()
 
 ###############################################################################
 ### WW3
```
Comment (aerorahul) [inline CMakeLists.txt:162]: and possibly WW3 too.
In the S2S model all these components are not optional.

File: CMakeLists.txt [hunk 6]
```diff
@@ -157,17 +238,17 @@ if(WW3)
...
-endif(WW3)
+endif()
 
 ###############################################################################
-### NEMS
+### NEMS (ufs-weather-model-lib)
 ###############################################################################
 add_library(
-    ufs-s2s-model-lib
+    ufs-weather-model-lib
```
Comment (aerorahul) [inline CMakeLists.txt]: this is a s2s model, not weather model

File: CMakeLists.txt [hunk 7]
```diff
@@ -181,25 +262,66 @@ add_library(
...
+    target_compile_definitions(ufs-weather-model-lib PRIVATE -DFRONT_CCPP)
+endif()
 
 if(WW3)
-    target_compile_definitions(ufs-s2s-model-lib PRIVATE -DFRONT_WW3=WMESMFMD)
-    target_include_directories(ufs-s2s-model-lib PRIVATE ${PROJECT_SOURCE_DIR}/WW3/model/mod)
-    set_target_properties(ufs-s2s-model-lib PROPERTIES Fortran_MODULE_DIRECTORY ${PROJECT_SOURCE_DIR}/WW3/model/mod)
-    add_dependencies(ufs-s2s-model-lib ww3_nems)
+    target_compile_definitions(ufs-weather-model-lib PRIVATE -DFRONT_WW3=WMESMFMD)
+    target_include_directories(ufs-weather-model-lib PRIVATE ${PROJECT_SOURCE_DIR}/WW3/model/mod)
```
Comment (aerorahul) [inline CMakeLists.txt]: PROJECT_SOURCE_DIR? Are these compiled modules in the source tree?
Shouldn't this be CMAKE_CURRENT_BINARY_DIR. Please check.

File: CMakeLists.txt [hunk 8]
```diff
@@ -181,25 +262,66 @@ add_library(
...
+
+if(MOM6)
+    target_compile_definitions(ufs-weather-model-lib PRIVATE -DFRONT_MOM6=mom_cap_mod)
+    target_include_directories(ufs-weather-model-lib PRIVATE ${CMAKE_CURRENT_BINARY_DIR}/MOM6-interface/mod)
+    set_target_properties(ufs-weather-model-lib PROPERTIES Fortran_MODULE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/MOM6-interface/mod)
+    add_dependencies(ufs-weather-model-lib nuopcmom6)
+    target_link_libraries(ufs-weather-model-lib PUBLIC nuopcmom6)
+    target_include_directories(ufs-weather-model-lib PUBLIC
+                                       $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/MOM6-interface/mod>)
+    list(APPEND optional_targets nuopcmom6)
```
Comment (aerorahul) [inline CMakeLists.txt]: this is not optional in S2S

File: CMakeLists.txt [hunk 9]
```diff
@@ -181,25 +262,66 @@ add_library(
...
+endif()
+
+if(CICE6)
```
Comment (aerorahul) [inline CMakeLists.txt]: CICE6 is not optional in S2S

File: CMakeLists.txt [hunk 10]
```diff
@@ -214,74 +336,46 @@ configure_file(${PROJECT_SOURCE_DIR}/NEMS/src/ESMFConvenienceMacros.h
...
 ###############################################################################
-add_executable(
-    NEMS.exe
-
-    NEMS/src/MAIN_NEMS.F90
+add_executable(NEMS.exe
+               NEMS/src/MAIN_NEMS.F90
 )
-add_dependencies(NEMS.exe ufs-s2s-model-lib)
+add_dependencies(NEMS.exe ufs-weather-model-lib fms)
```
Comment (aerorahul) [inline CMakeLists.txt]: something is wrong here.  the weather-model-lib depends on fms. and So there is no need for `fms` here.

File: CMakeLists.txt [hunk 11]
```diff
@@ -214,74 +336,46 @@ configure_file(${PROJECT_SOURCE_DIR}/NEMS/src/ESMFConvenienceMacros.h
...
+                      $ENV{PIO_LIBRARIES}/libpiof.a)
 
 ###############################################################################
 ### Install
 ###############################################################################
-
 install(
-  TARGETS ufs-s2s-model-lib fv3cap fv3core io ipd gfsphysics ${CCPP_LIBRARIES} fv3cpl stochastic_physics
-  EXPORT ufs-s2s-model-lib-config
+  TARGETS ufs-weather-model-lib fms ${optional_targets}
```
Comment (aerorahul) [inline CMakeLists.txt]: these are not optional targets.

File: fms_files.cmake [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    FMS/time_interp/time_interp.F90
+    FMS/time_interp/time_interp_external.F90
+    FMS/time_manager/get_cal_time.F90
+    FMS/time_manager/time_manager.F90
+    FMS/topography/gaussian_topog.F90
+    FMS/topography/topography.F90
+    FMS/tracer_manager/tracer_manager.F90
+    FMS/tridiagonal/tridiagonal.F90
+
+    FMS/affinity/affinity.c
```
Comment (aerorahul) [inline fms_files.cmake]: please remove the affinity.c file from this list.

**Review by aerorahul — CHANGES_REQUESTED** [review]
FIrst pass at the code.

File: CMakeLists.txt (hunk 6 above)
Comment (BrianCurtis-NOAA) [reply CMakeLists.txt]: This will be addressed in my fixes i push today.. I had caught that this morning.

File: fms_files.cmake (hunk 12 above)
Comment (BrianCurtis-NOAA) [reply fms_files.cmake]: This will be addressed in a commit coming.

File: CMakeLists.txt (hunk 2 above)
Comment (BrianCurtis-NOAA) [reply CMakeLists.txt]: Will change to if(S2S) that will enable CICE6/MOM6/CMEPS, then when -weather adopts CMEPS we can make it permanent later

File: CMakeLists.txt (hunk 3 above)
Comment (BrianCurtis-NOAA) [reply CMakeLists.txt]: Will change to if(S2S) that will enable CICE6/MOM6/CMEPS, then when -weather adopts CMEPS we can make it permanent later

File: CMakeLists.txt (hunk 4 above)
Comment (BrianCurtis-NOAA) [reply CMakeLists.txt]: Will change to if(S2S) that will enable CICE6/MOM6/CMEPS, then when -weather adopts CMEPS we can make it permanent later

File: CMakeLists.txt (hunk 5 above)
Comment (BrianCurtis-NOAA) [reply CMakeLists.txt:162]: WW3 is optional is S2S and Weather

File: CMakeLists.txt (hunk 8 above)
Comment (BrianCurtis-NOAA) [reply CMakeLists.txt]: I'll edit so that if(S2S) sets these as s2s_targets

File: CMakeLists.txt (hunk 9 above)
Comment (BrianCurtis-NOAA) [reply CMakeLists.txt]: Will change to if(S2S) that will enable CICE6/MOM6/CMEPS, then when -weather adopts CMEPS we can make it permanent later

File: CMakeLists.txt (hunk 7 above)
Comment (BrianCurtis-NOAA) [reply CMakeLists.txt]: The mod files for WW3 are in its source dir. Not sure why. We can change this later though.

**Review by BrianCurtis-NOAA — COMMENTED** [review]
These changes were made in more recent commits

**Review by aerorahul — CHANGES_REQUESTED** [review]
1. please remove `tests_cmake/` directory.
2. put all the changes in the `tests/` directory so we can compare what is actually changed.

The point here is to unify these, not create separate builds and logs.  `rt_s2s.conf` should not exist.  These changes should be in `tests/rt.conf`.

**Comment by BrianCurtis-NOAA** [comment]
@aerorahul Its only separated TEMPORARILY. The tests/ and parm/ directories between ufs-weather and ufs-s2s are drastically different (just look at the parm/ dir between the two). This separation is an effort to use the work done by Minsuk to unify the RT's between weather and s2s WITHOUT breaking the current S2S regression testing. If we mess with the current regression testing (just the tests/ dir) before submitting this PR then the workload added to keep up with all the RT changes would increase by a lot. We're trying to add to the S2S repo without breaking what it currently does, and your changes would not let that happen.

**Comment by JessicaMeixner-NOAA** [comment]
@BrianCurtis-NOAA at one point in time, the parm directory would be updated with updates from ufs-weather-model when we updated the atm model.  There were a few exceptions such as we have an extra variable in the model_configure (see: https://github.com/ufs-community/ufs-s2s-model/blob/develop/parm/model_configure.IN#L24) but I don't know if that's true or there has been further divergence more recently.

**Comment by aerorahul** [comment]
> @aerorahul Its only separated TEMPORARILY. The tests/ and parm/ directories between ufs-weather and ufs-s2s are drastically different (just look at the parm/ dir between the two). This separation is an effort to use the work done by Minsuk to unify the RT's between weather and s2s WITHOUT breaking the current S2S regression testing. If we mess with the current regression testing (just the tests/ dir) before submitting this PR then the workload added to keep up with all the RT changes would increase by a lot. We're trying to add to the S2S repo without breaking what it currently does, and your changes would not let that happen.

So, this directory `tests_cmake/` is not going to be merged into `develop`.

**Comment by BrianCurtis-NOAA** [comment]
OK. Lets keep tests_cmake out for now. This will be purely a PR to enable a CMake build of S2S and have NOTHING to do with testing.

**Comment by junwang-noaa** [comment]
We need to discuss the ufs-model test system further as we are going to
support several applications. In general, we need to keep consistency of
parm files, e.g. we only use parm files tested in standalone fv3 in
ufs-s2s. So,we don't use a fv3 namelist without being tested in
staloneslone as the ufs-s2s fv3 namelist.

I'd suggest to first commit the cmake build part to s2s, then make another
PR to confirm the RT giving same results as current gnu make when using
cmake. Tests _cmake is used to confirm the second step results.

**Comment by junwang-noaa** [comment]
> @BrianCurtis-NOAA at one point in time, the parm directory would be updated with updates from ufs-weather-model when we updated the atm model. There were a few exceptions such as we have an extra variable in the model_configure (see: https://github.com/ufs-community/ufs-s2s-model/blob/develop/parm/model_configure.IN#L24) but I don't know if that's true or there has been further divergence more recently.

I don't think we need that line though. We need to do some cleanup.

**Comment by DeniseWorthen** [comment]
Is there really some reason to think we don't need the line in mode_configure which contains the fast coupling frequency? That was a fix from a long time ago that allowed us to run the different slow-fast coupling timesteps. It is read in FV3/cpl/module_cap_cpl.F90.

**Comment by junwang-noaa** [comment]
Then we need to consolidate the files, set value -1 for ufs-weather, we
don't really need these two different versions of model_configure.IN for
ufs-weather and ufs-s2s stay under parm.

**Comment by BrianCurtis-NOAA** [comment]
Waiting on: https://github.com/NOAA-EMC/fv3atm/pull/183#event-3842002363 and https://github.com/noaa-psd/stochastic_physics/pull/30

**Pull request closed without merging**